Fix Half X86_F16 CUDA build failure

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141035
* __->__ #140994
* #140720
* #140567
* #140566
* #140565
* #140564

It passed PyTorch CI, but internally we saw failures from this.

Differential Revision: [D66137897](https://our.internmc.facebook.com/intern/diff/D66137897/)

swolchok committed:
Fix Half X86_F16 CUDA build failure

It passed PyTorch CI, but internally we saw failures from this.

Differential Revision: [D66137897](https://our.internmc.facebook.com/intern/diff/D66137897/)

[ghstack-poisoned]

swolchok commented:
@pytorchbot label "topic:not user facing"

malfet commented on c10/util/Half.h:
```diff
@@ -345,13 +345,9 @@ inline uint16_t fp16_ieee_from_fp32_value(float f) {
+#endif // C10_X86_F16
 }
 
-#ifdef X86_F16
-#undef X86_F16
-#endif // X86_F16
```
Why don't you want to undefine it at the end of the file?

swolchok replied:
I thought we were allowed to define macros as long as they were properly namespaced. Seemed unlikely to collide

Skylion007 replied:
Macros ignore C++ namespaces...

swolchok replied:
> Macros ignore C++ namespaces...

"C10_" is the namespace in my book.

swolchok replied:
anyway I updated the diff to undefine the macro

malfet reviewed (approved):
LGTM, sounds like you are addressing feedback left at https://github.com/pytorch/pytorch/pull/140720#discussion_r1847107150

swolchok committed:
Update on "Fix Half X86_F16 CUDA build failure"


It passed PyTorch CI, but internally we saw failures from this.

Differential Revision: [D66137897](https://our.internmc.facebook.com/intern/diff/D66137897/)

[ghstack-poisoned]

swolchok committed:
Update on "Fix Half X86_F16 CUDA build failure"


It passed PyTorch CI, but internally we saw failures from this.

Differential Revision: [D66137897](https://our.internmc.facebook.com/intern/diff/D66137897/)

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge

malfet commented:
Oops, this have landed already, but it would be nice to highlight, that it not only broke internal builds, but other ecosystem libraries, fbgemm for example, see https://github.com/pytorch/FBGEMM/pull/3394#issuecomment-2486370560
